core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java

```diff
@@ -108,37 +124,77 @@ void cleanupAllocatedDangledIndices(MetaData metaData) {
+            for (PersistedIndexMetaData persistedIndex : indexMetaDataList) {
+                if (persistedIndex != null) {
+                    final Index index = persistedIndex.getIndexMetaData().getIndex();
+                    final String clusterUUID = metaData.clusterUUID();
+                    if (persistedIndex.isSameClusterUUID(clusterUUID)) {
+                        // the cluster metadata's cluster UUID is the same as the cluster UUID that created this index,
+                        // so the index shouldn't be imported as dangling but rather deleted and cleaned up, because the
+                        // dangling files are probably due to remnants from deleted indices (e.g. the node was offline while
+                        // the index was deleted) or failed cleanup
+                        logger.info("[{}] dangling index exists on the local file system, but it is deleted in the metadata of the " +
+                                    "same cluster that created it, deleting the index on the local file system", index.getName());
+                        delIndices.put(index, persistedIndex.getIndexMetaData());
+                    } else {
+                        if (metaData.hasIndex(index.getName())) {
+                            logger.warn("[{}] can not be imported as a dangling index, as index with same name already exists " +
+                                        "in cluster metadata", index.getName());
+                        } else {
+                            logger.info("[{}] dangling index, exists on local file system, but not in cluster metadata, " +
+                                        "auto import to cluster state", index.getName());
+                            newIndices.put(index, persistedIndex.getIndexMetaData());
+                        }
+                    }
                 }
             }
-            return newIndices;
+            return Tuple.tuple(newIndices, delIndices);
         } catch (IOException e) {
             logger.warn("failed to list dangling indices", e);
-            return emptyMap();
+            return Tuple.tuple(emptyMap(), emptyMap());
+        }
+    }
+
+    /**
+     * Cleans the indices that exist on the local file system but should be deleted according to the metadata state
+     * and we have decided that the indices should not be imported as dangling.
+     */
+    void cleanDeletedIndices() {
+        if (deletedIndices.isEmpty()) {
+            return;
+        }
+        for (Iterator<Map.Entry<Index, IndexMetaData>> entries = deletedIndices.entrySet().iterator(); entries.hasNext();) {
+            Map.Entry<Index, IndexMetaData> entry = entries.next();
+            final String indexName = entry.getKey().getName();
+            final IndexMetaData indexMetaData = entry.getValue();
+            try {
+                nodeEnv.deleteIndexDirectoryUnderLock(indexMetaData.getIndex(), new IndexSettings(indexMetaData, settings));
```

Comment:
Oh oh! deleteUnderLock should be called when you hold the lock! instead we should use IndicesService.deleteIndexStore